Documentation/tectonic.md

```diff
@@ -0,0 +1,52 @@
+# Rook on Tectonic
+
+Here is a running guide on how to implement Rook on Tectonic. A complete guide on how to install Tectonic is out of the scope of the Rook project. But many info can be found on the [Tectonic website](https://coreos.com/tectonic/docs/latest/)
+
+## Prequisites
+
+- An installed tectonic-installer. These steps are described on [the Tectonic website](https://coreos.com/tectonic/docs/latest/install/bare-metal/#4-tectonic-installer)
+- A running matchbox node which will do the provisioning (Bare metal)
+- You can run through all steps of the GUI installer, but in the last step, choose `Boot manually`. This way we can make the necessary changes first.
+
+## Edit the kubelet.service file to be compatible with Rook
+
+To make the kubelet compatible with Rook we need to make a few adaptions to the systemd service file generated by the Tectonic-installer.
+
+First cd into the directory where you untarred the tectonic installer and find your newly generated cluster configuration files.
+
+```
+ $ cd ~/tectonic/tectonic-installer/LINUX-OR-DARWIN/clusters
+ $ nano modules/ignition/resources/services/kubelet.service
+```
+
+And after the line `ExecStartPre=/bin/mkdir -p /var/lib/cni`, paste the following extra lines:
+
+```
+ExecStartPre=/bin/mkdir -p /var/lib/kubelet/volumeplugins
+ExecStartPre=/bin/mkdir -p /var/lib/rook
+```
+
+And a little down in the `ExecStart` section, add the following kubelet option:
+
+```
+--volume-plugin-dir=/var/lib/kubelet/volumeplugins \
+```
+
+Save the file and close.
+
+### Provision your matchbox
+
+Still from this cluster directory, invoke the following commands:
+
+```
+ $ terraform plan -var-file=terraform.tfvars platforms/metal
+ $ terraform apply -var-file=terraform.tfvars platforms/metal
+```
+
+Now if all went well the correct files were automatically provisioned to your Matchbox node and you can boot all your cluster nodes. As an experience from XenServer this iPXE boot is really slow, so you can go to grab something to eat by now. Changes are big that the `terraform apply` command will fail after one hour, but just restart the same command again to finish up.
+
+## Start Rook
+
+Follow the instructions as stated in [the installation docs](https://rook.github.io/docs/rook/master/kubernetes.html). Best thing is to provide the disks to use in the [rook-cluster.yaml](https://rook.github.io/docs/rook/master/cluster-crd.html#storage-configuration-settings)
+
+TODO:: Integrate the prometheus with the prometheus from Tectonic
```

Comment:
How about you link to the example ServiceMonitor in `cluster/examples/kubernetes/`?
Users can then look at the ServiceMonitor and change it as required. Additionally you could add a hint here for "what to change in the ServiceMonitor" or in "which namespace to create it", etc for it to work with the Prometheus in Tectonic.

Reply:
I tried for some time to get Rook implemented in the Tectonic installation. So far no success, the prometheus monitor doesn't appear in the Prometheus -> Targets list yet.
A few things I already discovered:
- Prometheus object filters on label  `k8s-app` which must exist.
- Added namespaceSelector (rook) to the  selector object to make sure it points to the rook namespace.

A possible problem is the RBAC (I'm not yet very familiar with this)

Currently my `service-monitor-tectonic.yaml` file looks like

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: rook-api
  namespace: tectonic-system
  labels:
    team: rook
    k8s-app: rook
spec:
  selector:
    namespaceSelector:
      matchNames:
      - rook
    matchLabels:
      app: rook-api
      rook_cluster: rook
  endpoints:
  - port: rook-api
    targetPort: 0
    interval: 5s
```

I'll let you know when I have more progress

Reply:
@galexrt I found the solution and added two files in the `cluster/example/kubernetes/monitoring` directory.
- `tectonic-prometheus.yaml` which adds the roles for the Tectonic prometheus-k8s service account
- `tectonic-service-monitor.yaml` which creates the service monitor for Rook

In which file do you suggest I create the creation for Tectonic monitoring files?
```bash
kubectl create -f tectonic-prometheus.yaml
kubectl create -f tectonic-service-monitor.yaml
```
in `tectonic.md` or with a new header inside `k8s-monitoring.md`?

`tectonic-prometheus.yaml`
```
kind: ClusterRole
metadata:
  name: prometheus-rook
  namespace: rook
rules:
- apiGroups: [""]
  resources:
  - nodes
  - services
  - endpoints
  - pods
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["get"]
- nonResourceURLs: ["/metrics"]
  verbs: ["get"]
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: prometheus-rook
  namespace: rook
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: prometheus-rook
subjects:
- kind: ServiceAccount
  name: prometheus-k8s
  namespace: tectonic-system
```

`tectonic-service-monitor.yaml`

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: rook
  namespace: tectonic-system
  labels:
    team: rook
    k8s-app: rook
spec:
  namespaceSelector:
    matchNames:
    - rook
  selector:
    matchLabels:
      app: rook-api
      rook_cluster: rook
  endpoints:
  - port: rook-api
    targetPort: 8124
    path: /metrics
    interval: 5s
```

I had  to add `targetPort: 8124` which is not ideal, but it was needed to get it working. This might be because I use calico for network policies and created Rook with the option 'hostNetwork: true` to avoid data going through calico network policies all the time

Let me know what you think and I'll update the pull request

Reply:
The ServiceMonitor seems to work, although adding the [Ceph-cluster](https://grafana.com/dashboards/917) grafana dashboard seemed more complicated. It always got deleted automatically after a minute or so.

After having communicated in [this issue](https://github.com/coreos/prometheus-operator/issues/772) with the Tectonic/CoreOS folks, it seems they rather have Rook running a separate Prometheus cluster, so they can handle automatic updates on their Kubernetes Prometheus, Grafana and  Alertmanager pods.

I will create all the necessary manifests and update the pull request!